File: src/42_Trapping_Rain_Water.rs
// 42_Trapping_Rain_Water
// https://leetcode.cn/problems/trapping-rain-water/


// 可惜超时
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let &highst = height.iter().max().unwrap();
        let len = height.len();
        let mut wall = vec![vec![0; len]; highst as usize];
        for i in 0..highst as usize {
            for j in 0..len {
                if highst - i as i32 <= height[j] {
                    wall[i][j] = 1;
                }
            }
        }
        
        let mut ans = 0;
        for i in 0..highst as usize {
            let mut left = 0;
            let mut right = len-1;
            while left < right && wall[i][left] != 1 {
                left += 1;
            }
            while right > left && wall[i][right] != 1 {
                right -= 1;
            }
            for j in left+1..right {
                if wall[i][j] == 0 {
                    ans += 1;
                }
            }
        }
    
        ans
    }
}

/// 单调栈
/// 还是不熟悉，看着思路都没写出来....
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let mut stk: Vec<usize>= vec![];
        let mut ans = 0;

        for i in 0..height.len() {
            while !stk.is_empty() && height[stk[stk.len()-1]] < height[i] {
                let top = stk.pop().unwrap();
                if stk.is_empty() {
                    break;
                }
                let left = stk[stk.len()-1];
                ans += (i-left-1)*((height[i].min(height[left])-height[top]) as usize);
            }
            stk.push(i);
        }

        ans as i32
    }
}

File: src/42_Trapping_Rain_Water.py
# 42_Trapping_Rain_Water
# https://leetcode.cn/problems/trapping-rain-water/

# 给定 n 个非负整数表示每个宽度为 1 的柱子的高度图，计算按此排列的柱子，下雨之后能接多少雨水。

# 输入：height = [0,1,0,2,1,0,1,3,2,1,2,1]
# 输出：6
# 解释：上面是由数组 [0,1,0,2,1,0,1,3,2,1,2,1] 表示的高度图，在这种情况下，可以接 6 个单位的雨水（蓝色部分表示雨水）。 

from typing import List
class Solution:
    def trap(self, height: List[int]) -> int:
        stack = []
        ans = 0
        for i, h in enumerate(height):
            if stack is None:
                stack.append([i,h])
                continue
            while len(stack) and h > stack[-1][1]:
                pop_first = stack.pop()
                if len(stack) >= 1:
                    cur_peak = stack[-1]
                    l = min(cur_peak[1], h) - pop_first[1]
                    r = i - cur_peak[0] - 1
                    ans += l * r
            stack.append([i,h])

        return ans
            
            
solution = Solution()
print(solution.trap(height = [0,1,0,2,1,0,1,3,2,1,2,1]))